scripts/test_solidity_lsp.py

```diff
@@ -0,0 +1,379 @@
+        if message['method'] != method_name:
+            return None
+        return message['params']
+
+    def open_file_and_wait_for_diagnostics(self,
+                                           solc: JsonRpcProcess,
+                                           test_case_name: str,
+                                           diagnostic_reports: int = 1) -> List[Any]:
+        """
+        Opens file for given test case and waits for diagnostics to be published.
+        """
+        reports = []
+        reply = solc.call_method('textDocument/didOpen',
+            {
+                'textDocument': {
+                    'uri': self.get_test_file_uri(test_case_name),
+                    'languageId': 'Solidity',
+                    'version': 1,
+                    'text': self.get_test_file_contents(test_case_name)
+                }
+            }
+        )
+        diags = self.extract_params_for_method('textDocument/publishDiagnostics', reply)
+        if diags == None:
+            raise RuntimeError(f"Unepxected response from RPC endpoint: {reply}")
+        reports.append(diags)
+        # reply now contains one "textDocument/publishDiagnostics" notification
+        while diagnostic_reports > 1:
+            diagnostic_reports -= 1
+            reply = solc.receive_message()
+            diags = self.extract_params_for_method('textDocument/publishDiagnostics', reply)
+            if diags == None:
+                raise RuntimeError(f"Unepxected response from RPC endpoint: {reply}")
+            reports.append(diags)
+        return reports
+
+    def expect_equal(self, actual, expected, description="Equality") -> None:
+        self.assertion_count = self.assertion_count + 1
+        prefix = f"[{self.assertion_count}] {SGR_ASSERT_BEGIN}{description}: "
+        diff = DeepDiff(actual, expected)
+        if len(diff) == 0:
+            self.assertions_passed = self.assertions_passed + 1
+            if self.print_assertions:
+                print(prefix + SGR_STATUS_OKAY + 'OK' + SGR_RESET)
+            return
+
+        # Failed assertions are always printed.
+        self.assertions_failed = self.assertions_failed + 1
+        print(prefix + SGR_STATUS_FAIL + 'FAILED' + SGR_RESET)
+        raise ExpectationFailed(actual, expected)
+
+    # pylint: disable-next=too-many-arguments
+    def expect_diagnostic(self, diagnostic, code: int, lineNo: int, startColumn: int, endColumn: int):
+        self.expect_equal(diagnostic['code'], code, f'diagnostic: {code}')
+        self.expect_equal(
+            diagnostic['range'],
+            {'end': {'character': endColumn, 'line': lineNo},
+             'start': {'character': startColumn, 'line': lineNo}},
+            "diagnostic: check range"
+        )
+    # }}}
+
+    # {{{ actual tests
+    def test_publish_diagnostics_1(self, solc: JsonRpcProcess) -> None:
+        """ Publish diagnostics (warnings) """
+        self.setup_lsp(solc)
+        TEST_NAME = 'publish_diagnostics_1'
```

Comment:
Why not just call one test `publish_diagnostics_warnings` and the other `publish_diagnostics_errors` Instead of numbering them? 

BTW I really wish our test names were more verbose because checking if we already have a test covering something just by looking at file names is not easy.

Reply:
That's already done btw, I thought I did push it. Ugh oh.